tests/tas-elasticsearch/src/test/java/org/alfresco/elasticsearch/upgrade/AvailabilityProbe.java

```diff
@@ -0,0 +1,135 @@
+package org.alfresco.elasticsearch.upgrade;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Supplier;
+
+import com.google.common.util.concurrent.AtomicLongMap;
+import com.google.common.util.concurrent.RateLimiter;
+import com.google.common.util.concurrent.Uninterruptibles;
+
+class AvailabilityProbe
+{
+    private final Thread thread;
+    private final RateLimiter rateLimiter;
+    private final AtomicBoolean stopRequested = new AtomicBoolean();
+    private final Supplier<ProbeResult> probingFunction;
+    private final AtomicLongMap<ProbeResult> stats = AtomicLongMap.create();
+
+    public static AvailabilityProbe createRunning(int requestsPerSecond, Supplier<ProbeResult> probingFunction)
+    {
+        final AvailabilityProbe probe = new AvailabilityProbe(requestsPerSecond, probingFunction);
+        probe.start();
+        return probe;
+    }
+
+    private AvailabilityProbe(int requestsPerSecond, Supplier<ProbeResult> probingFunction)
+    {
+        rateLimiter = RateLimiter.create(requestsPerSecond);
+        this.probingFunction = probingFunction;
+        thread = new Thread(this::probing);
+    }
+
+    public Stats getStats()
+    {
+        return new Stats(stats.asMap());
+    }
+
+    public Stats stop()
+    {
+        stopRequested.set(true);
+        Uninterruptibles.joinUninterruptibly(thread, 10, TimeUnit.SECONDS);
+        return getStats();
+    }
+
+    private void start()
+    {
+        if (thread.isAlive()) return;
+        thread.start();
+    }
+
+    private void probing()
+    {
+        while (!stopRequested.get())
```

Comment:
I don't think I have enough context, but is there a possibility of locking up an additional thread running in endless loop if an exception occurs before "stop()" method is called?

Reply:
Yes it is. This is why stop method should be called ;)

Reply:
I changed it to a daemon thread so it won't block the jvm shutdown when it's not stopped correctly.

Reply:
I introduced a change to stop the availability probe thread along with the environment it is running on.